docs/fundamentals/database-collection.txt

```diff
@@ -0,0 +1,176 @@
+.. _laravel-db-coll:
+
+=========================
+Databases and Collections
+=========================
+
+.. facet::
+   :name: genre
+   :values: reference
+
+.. meta::
+   :keywords: php framework, odm
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to use {+odm-short+} to access
+and manage MongoDB databases and collections.
+
+MongoDB organizes data in a hierarchical structure. A MongoDB
+deployment contains one or more **databases**, and each database
+contains one or more **collections**. In each collection, MongoDB stores
+data as **documents** that contain field-and-value pairs. In
+{+odm-short+}, you can access documents through Eloquent models.
+
+To learn more about the document data format,
+see :manual:`Documents </core/document/>` in the Server manual.
+
+.. _laravel-access-db:
+
+Specify the Database in a Connection Configuration
+--------------------------------------------------
+
+You specify a database name when configuring a connection in your
+application's ``config/database.php`` file. The ``connections`` property
+in this file stores all of your database connection information, such as
+your connection string, database name, and optionally, authentication
+details. After you specify a database connection, you can perform
+database-level operations and access collections that the database
+contains.
+
+If you set the database name in a connection string to the name of a
+nonexistent database, Laravel still makes a valid connection. When you
+insert any data into a collection in the database, the server creates it
+automatically.
+
+The following example shows how to set a default database connection and
+create a database connection to the ``animals`` database in the
+``config/database.php`` file:
+
+.. code-block:: php
+   :emphasize-lines: 7
+   
+   'default' => 'mongodb',
+
+   'connections' => [
+
+       'mongodb' => [
+           'driver' => 'mongodb',
+           'dsn' => 'mongodb://localhost:27017/animals',
+       ], ...
+   ]
+
+When you set a default database connection, {+odm-short+} uses that
+connection for operations, but you can specify multiple database connections
+in your ``config/database.php`` file.
+
+The following example shows how to specify multiple database connections
+(``mongodb`` and ``mongodb_alt``) to access the ``animals`` and
+``plants`` databases:
```

Comment:
@GromNaN: Do we really not have any alternative way to utilize multiple databases with this library? It's bizarre to me that separate DSNs would be required, and that we'd have to rely on the deprecated auth database component of connection string.

If users need to actually use a DSN with authentication, and omit `authSource`, they'll be in for a bad time.

In Doctrine, we have the ability to use a single connection and just specify different databases as needed for models.

This example is admittedly contrived, since "animals" and "plants" sound like they're used by the same application, and a single application should generally use one database (excluding perhaps something like extra logging).

Reply:
Just re-read the class `Laravel\MongoDB\Connection`.

- When the `dsn` config is not set, it is built from the config `host` (string or array) with the auth database that can be read from the config **option** `database` (different from the `database` config in the config). https://github.com/mongodb/laravel-mongodb/blob/bd6aa40bdbd285d9318a20b7dff6b96270115e75/src/Connection.php#L252-L255
Check de examples: https://github.com/mongodb/laravel-mongodb/blob/bd6aa40bdbd285d9318a20b7dff6b96270115e75/tests/ConnectionTest.php#L50-L170
- The database name is read from the `database` config. If not set, the database name is extracted from the dsn, after the `/`. https://github.com/mongodb/laravel-mongodb/blob/bd6aa40bdbd285d9318a20b7dff6b96270115e75/src/Connection.php#L155-L163
- Any additionnal option for the client can be set in `options`: `username`, `password`, `database` for auth database, and all the [driver options](https://www.mongodb.com/docs/php-library/current/reference/method/MongoDBClient__construct/#parameters).

Reply:
@GromNaN: I can't make sense of everything in that comment, but the ["DSN takes precedence over host/port config" test case](https://github.com/mongodb/laravel-mongodb/blob/bd6aa40bdbd285d9318a20b7dff6b96270115e75/tests/ConnectionTest.php#L155) stood out to me. There, a `config.database` option appears to take precedence over the auth database (i.e. path component of the URI) in the `config.dsn` option.

Given that, it seems like there's absolutely no reason we need to specify the "default" database name in a URI, and risk interfering with connection persistence.